tests/AnalyzeProperties.hs

```diff
@@ -0,0 +1,575 @@
+mkDec = pure . ESimple TDecimal . Inj . Numerical
+
+class Extract a where
+  extract :: ETerm -> Analyze.Term a
+
+instance Extract Integer where
+  extract = \case
+    ESimple TInt x -> x
+    other -> error (show other)
+
+instance Extract Decimal where
+  extract = \case
+    ESimple TDecimal x -> x
+    other -> error (show other)
+
+instance Extract String where
+  extract = \case
+    ESimple TStr x -> x
+    other -> error (show other)
+
+instance Extract Bool where
+  extract = \case
+    ESimple TBool x -> x
+    other -> error (show other)
+
+instance Extract Time where
+  extract = \case
+    ESimple TTime x -> x
+    other -> error (show other)
+
+-- TODO: Var, objects
+-- TODO: we might want to reweight these by using `Gen.frequency`.
+genCore :: MonadGen m => SizedType -> m ETerm
+genCore (SizedInt size) = Gen.recursive Gen.choice [
+    ESimple TInt . CoreTerm . Lit <$> genInteger size
+  ] [
+    Gen.subtermM2 (genCore (SizedInt size)) (genCore (SizedInt (1 ... 1e3))) $
+      \x y -> mkInt $ Numerical $ ModOp (extract x) (extract y)
+  , do op <- genArithOp
+       let (size1, size2) = arithSize op size
+       Gen.subtermM2 (genCore (SizedInt size1)) (genCore (SizedInt size2)) $
+         \x y -> mkInt $ Numerical $ IntArithOp op (extract x) (extract y)
+  , do op <- genUnaryArithOp
+       let size' = unaryArithSize op size
+       Gen.subtermM (genCore (SizedInt size')) $
+         mkInt . Numerical . IntUnaryArithOp op . extract
+  , Gen.subtermM (genCore (SizedDecimal size)) $ \x -> do
+    op <- genRoundingLikeOp
+    mkInt $ Numerical $ RoundingLikeOp1 op (extract x)
+  , Gen.subtermM (genCore strSize) $ mkInt . StrLength . extract
+  ]
+genCore sizeD@(SizedDecimal size) = Gen.recursive Gen.choice [
+    ESimple TDecimal . CoreTerm . Lit <$> genDecimal size
+  ] [
+    do op <- genArithOp
+       let (size1, size2) = arithSize op size
+       Gen.subtermM2 (genCore (SizedDecimal size1)) (genCore (SizedDecimal size2)) $
+         \x y -> mkDec $ DecArithOp op (extract x) (extract y)
+  , do
+       op <- genUnaryArithOp
+       let size' = unaryArithSize op size
+       Gen.subtermM (genCore (SizedDecimal size')) $
+         mkDec . DecUnaryArithOp op . extract
+  , do op <- genArithOp
+       let (size1, size2) = arithSize op size
+       Gen.subtermM2 (genCore (SizedDecimal size1)) (genCore (SizedInt size2)) $
+         \x y -> mkDec $ DecIntArithOp op (extract x) (extract y)
+  , do
+       op <- genArithOp
+       let (size1, size2) = arithSize op size
+       Gen.subtermM2 (genCore (SizedInt size1)) (genCore (SizedDecimal size2)) $
+         \x y -> mkDec $ IntDecArithOp op (extract x) (extract y)
+  , Gen.subtermM2 (genCore sizeD) (genCore (SizedInt (0 +/- 255))) $ \x y -> do
+      op <- genRoundingLikeOp
+      mkDec $ RoundingLikeOp2 op (extract x) (extract y)
+  ]
+genCore (SizedString len) = Gen.recursive Gen.choice [
+    -- TODO: use unicodeAll?
```

Comment:
`unicode`: Generates a Unicode character, excluding noncharacters and invalid standalone surrogates: `'\0'..'\1114111'` (excluding `'\55296'..'\57343'`)

`unicodeAll`: Generates a Unicode character, including noncharacters and invalid standalone surrogates: `'\0'..'\1114111'`

Our tests pass with `unicode` and fail with `unicodeAll`. I haven't decided yet if this is significant.

Reply:
`unicode` seems like it would be reasonable if e.g. we knew that validators would deny transactions that contain invalid UTF strings in Pact. @slpopejoy?

Reply:
this seems like the kind of thing we should make sure doesn't get past the pact parser